### PR TITLE
DR2-1963 Only one content folder per consignment

### PR DESCRIPTION
**Pull request by MancunianSam**
This is quite tricky to do without having to parse the file twice, once
for the unique consignments and then again to create the metadata file.

Instead, there's a AtomicCell with an map which is updated when we see
a particular consignment.

### CODE REVIEW

File: scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala [hunk 1]
```diff
@@ -162,6 +162,67 @@ class LambdaTest extends AnyFlatSpec with ScalaCheckDrivenPropertyChecks:
...
+    val input = Input(groupId, batchId, 1, 2)
+    val (s3Contents, output) = runHandler(uuidIterator, initialS3Objects, initialDynamoObjects, input)
+
+    val metadataObjects: List[MetadataObject] = s3Contents(s"$batchId/metadata.json").asInstanceOf[List[MetadataObject]]
+    val assetMetadataObjects = metadataObjects.collect { case assetMetadataObject: AssetMetadataObject => assetMetadataObject }
+    val contentFolderMetadataObjects = metadataObjects.collect { case contentFolderMetadataObject: ContentFolderMetadataObject => contentFolderMetadataObject }
+
+    contentFolderMetadataObjects.size should equal(2)
+
+    assetMetadataObjects.count(_.parentId == Option(contentFolderMetadataObjects.head.id)) should equal(10)
```
Comment (techncl) [inline scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala:221]: Do you think it's worth checking that the correct files went into the correct Content folders?

**Review by techncl — COMMENTED** [review]
Looks good! Just a few comments. Also, should we update the README with the fact that it's one content folder per consignment?

File: scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala [hunk 2]
```diff
@@ -79,8 +84,16 @@ class Lambda extends LambdaRunner[Input, Output, Config, Dependencies]:
                   fileLocation,
                   tdrMetadata.SHA256ServerSideChecksum
                 )
-                List(contentFolderMetadata, assetMetadata, fileMetadata)
+                val contentFolder = contentFolderMap.get(tdrMetadata.ConsignmentReference)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:87]: `potentialContentFolder` is ok. `potentialContentFolderMetadata` is getting a bit long. I've changed `contentFolderMetadata` on line 91 to match `potentialContentFolder` Arguably the case class names don't need `MetadataObject` in them so I can live with it being called `contentFolderMetadata`

File: scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala [hunk 3]
```diff
@@ -79,8 +84,16 @@ class Lambda extends LambdaRunner[Input, Output, Config, Dependencies]:
...
+                if contentFolder.isDefined then (contentFolderMap, List(assetMetadata.copy(parentId = contentFolder.map(_.id)), fileMetadata))
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:88]: I'm going to leave it as is. Yours is shorter but I quite like having the seperation in the `if` and `else` blocks between existing content folder and new one. 

I've got rid of the tuple brackets though, thanks.

File: scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala [hunk 4]
```diff
@@ -64,9 +68,10 @@ class Lambda extends LambdaRunner[Input, Output, Config, Dependencies]:
             )
           )
           Stream.evals {
-            dependencies.s3Client
-              .headObject(fileLocation.getHost, fileLocation.getPath.drop(1))
-              .map { headObjectResponse =>
+            for {
+              headObjectResponse <- dependencies.s3Client
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:72]: It's the response from `headObject` with a type of `HeadObjectResponse` so I think it's ok. File is implied as s3 is file storage.

File: scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala [hunk 5]
```diff
@@ -64,9 +68,10 @@ class Lambda extends LambdaRunner[Input, Output, Config, Dependencies]:
...
+                .headObject(fileLocation.getHost, fileLocation.getPath.drop(1))
+              res <- contentFolderCell.modify[List[MetadataObject]] { contentFolderMap =>
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:74]: `childObjects` is good.

File: scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala [hunk 6]
```diff
@@ -162,6 +162,67 @@ class LambdaTest extends AnyFlatSpec with ScalaCheckDrivenPropertyChecks:
...
+      None,
+      "body",
+      "2024-10-18 00:00:01",
+      "TDR-ABCD",
+      s"file.txt",
+      "checksum",
+      "reference"
+    )
+
+    val tdrMetadataTwo = TDRMetadata(
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala:184]: Good idea, thanks.

File: scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala (hunk 1 above)
Comment (MancunianSam) [reply scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala:221]: But we are checking that. We're expecting 10 with a parent of one content folder and 10 with the parent of another which is what the check does. We're not bothered what the ids are so long as they're in the correct structure.

Comment (techncl) [reply scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/LambdaTest.scala:211]: Ah! Sorry, I misread this 🤦

**Review by techncl — APPROVED** [review]
👍

**Pull request merged**